sl_server.cc

```diff
@@ -0,0 +1,353 @@
+      for (int j = 0; j < followers.followers(i).follows_size(); j++) {
+        if (followers.followers(i).follows(j) == request->username()) {
+          add_follow = false;
+        }
+      }
+      if (add_follow) {
+        (followers.mutable_followers(i))->add_follows(request->username());
+      }
+    }
+  }
+
+  // put new lists in database
+  following.SerializeToString(&following_serial);
+  followers.SerializeToString(&followers_serial);
+
+  client_.put("following", following_serial);
+  client_.put("followers", followers_serial);
+  mtx_.unlock();
+  return Status::OK;
+}
+
+Status SL_Server::read(ServerContext* context, const ReadRequest* request,
+                ReadReply* reply){
+  mtx_.lock();
+
+  // check if valid chirp id is provided
+  std::string chirps_serial = client_.get("chirps");
+  Chirps chirps;
+  chirps.ParseFromString(chirps_serial);
+  try {
+    if((std::stoi(request->chirp_id()) < 0) || (std::stoi(request->chirp_id()) >= chirps.chirps_size())) {
+      mtx_.unlock();
+      return Status(StatusCode::INVALID_ARGUMENT, "invalid chirp id");
+    }
+  } catch (std::invalid_argument&) {
+    mtx_.unlock();
+    return Status(StatusCode::INVALID_ARGUMENT, "invalid chirp id");
+  }
+
+  // get all replies from database
+  std::string replies_serial = client_.get("replies");
+  Replies replies;
+  replies.ParseFromString(replies_serial);
+
+  // convert replies to vector of vector
+  std::vector<std::vector<int> > replies_vec;
+  for(int i = 0; i < replies.allreplies_size(); i++) {
+    std::vector<int> chirp_replies;
+    for(int j = 0; j < replies.allreplies(i).replies_size(); j++) {
+      chirp_replies.push_back(replies.allreplies(i).replies(j));
+    }
+    replies_vec.push_back(chirp_replies);
+  }
+  //implement DFS to display all read chirps
+  std::vector<bool> visited(chirps.chirps().size(), false);
+  std::stack<int> dfs_stack;
+  dfs_stack.push(std::stoi(request->chirp_id()));
+  Chirp *chirp = reply->add_chirps();
+  *chirp = chirps.chirps(std::stoi(request->chirp_id()));
+  visited[std::stoi(request->chirp_id())] = true;
+  while(!dfs_stack.empty()) {
+    int curr_id = dfs_stack.top();
+    bool child_exists = false;
+    for(int i = 0; i < replies_vec[curr_id].size(); i++){
+      int reply_id = replies_vec[curr_id][i];
+      if(!visited[reply_id]) {
+        Chirp *chirpReply = reply->add_chirps();
+        *chirpReply = chirps.chirps(reply_id);
+        visited[reply_id] = true;
+        dfs_stack.push(reply_id);
+        child_exists = true;
+        break;
+      }
+    }
+    if(!child_exists) {
+      dfs_stack.pop();
+    }
+  }
+
+  mtx_.unlock();
+  return Status::OK;
+}
+
+Status SL_Server::monitor(ServerContext* context, const MonitorRequest* request, ServerWriter<MonitorReply>* writer){
+  // find list of chirps to stream from
+  std::string monitor_key = "monitor::" + request->username();
+  while(!context->IsCancelled()) {
+    mtx_.lock();
```

Comment:
isn't this spinlocking lol @our OS midterm today. doesn't this take up a lot of resources unnecessarily? idk if it's that important but it reminded me how we didn't know what was going on before when we first wrote code.

Reply:
It is wasteful, but my understanding was that he wanted us to poll the server to get the chirps when the user is monitoring.